client/network/src/behaviour.rs

```diff
@@ -172,6 +174,23 @@ pub enum BehaviourOut<B: BlockT> {
 	/// Events generated by a DHT as a response to get_value or put_value requests as well as the
 	/// request duration.
 	Dht(DhtEvent, Duration),
+
+	/// Event generated by bitswap.
+	Bitswap(BitswapEvent)
+}
+
+/// An event generated by bitswap.
+#[derive(Clone, Debug)]
+pub enum BitswapEvent {
+	/// A block has been received.
+	///
+	/// There are no guarantees about the recieved block:
+	/// * The CID and block contents might not match.
+	/// * We might have receieved a block without having sent out a want request first.
```

Comment:
That sounds like an attack vector.
@dvc94ch Can a node just connect and send, say, 2 GiB of data without us having asked for it, and we will simply buffer it?

Reply:
This should probably be combined with the reputation system right?

Reply:
no this can't happen https://github.com/ipfs-rust/tiny-multihash/issues/12

Reply:
@dvc94ch right, but that 2 GiB of data is still held in memory for a period of time, right?

Reply:
Well that depends on the transport. A message is processed from the queue, and dropped if it wasn't requested. We don't have a sophisticated mechanism for disconnecting from peers, I thought that substrate handles bandwidth limiting, telemetry and reputation, and instead of handling this stuff myself in bitswap/ipfs-embed, I thought we can leverage that. I'm glad to redesign bitswap if there is se input on how it should work

Reply:
FYI didn't even see the other comment on cid and data don't match. The bitswap never sends a hash, only the data. The cid is reconstructed while parsing the message, they are guaranteed to match https://github.com/ipfs-rust/libp2p-bitswap/blob/23f67ea8e36230aac4ef5c1feef51af6f157df38/src/message.rs#L155

Reply:
>  A message is processed from the queue, and dropped if it wasn't requested. We don't have a sophisticated mechanism for disconnecting from peers, I thought that substrate handles bandwidth limiting, telemetry and reputation, and instead of handling this stuff myself in bitswap/ipfs-embed, I thought we can leverage that. I'm glad to redesign bitswap if there is some input on how it should work

I'm not familiar with the bitswap protocol or the internals of bitswap, so I'm just naively asking.
The problem I have in mind is: someone can just connect, open 1000 substreams (or whatever the maximum is), negotiate bitswap, send a length prefix of `N`, and then send `N - 1` bytes but never the last byte. Do this multiple times simultaneously and you can probably make the node run out of memory.
This is entirely specific to the bitswap protocol, and not something Substrate can handle.

What we do for other protocols is enforce a limit of one substream per protocol per connection, and make sure that the remote can never send a large amount of data without having explicitly requested it first.

Reply:
> What we do for other protocols is enforce a limit of one substream per protocol per connection

https://github.com/ipfs-rust/libp2p-bitswap/issues/7

> make sure that the remote can never send a large amount of data without having explicitly requested it first.

I think this is already handled. Does this satisfy the requirement? https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/behaviour.rs#L216 Or is more work needed?

Reply:
> can you point me to a specific protocol that handles this, so I can look how it's done?

I'm not familiar with this code unfortunately, but [libp2p's request-response should do that](https://github.com/libp2p/rust-libp2p/tree/master/protocols/request-response/src)

> I think this is already handled. Does this satisfy the requirement? https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/behaviour.rs#L216 Or is more work needed?

As far as I understand, you're still entirely buffering the message, and then only discard it, so it really depends on the maximum size of the message. If it's 512 kiB it's ok, if it's 2 GiB it's not.

Reply:
I think these issues are handled much better now https://github.com/ipfs-rust/libp2p-bitswap/pull/8.

@tomaka how concerned are you with interopability with other implementations? In particular now we avoid packing multiple requests responses in the same message, and reuse the substream to send multiple messages. Also even if we add that supporting bitswap 1.1.0 and 1.0.0 is not something I'm particularly interested in working on, and the api is really focused around bitswap 1.2.0 which works completely different from previous versions.

Reply:
> how concerned are you with interopability with other implementations?

If this feature is used specifically for large files storage using Substrate, I guess we don't really care.